### PR TITLE
Group4 rpc impl

### PR DESCRIPTION
**Pull request by pelmers**
@Swyrough can you check that I didn't lose anything in the merge? it seems to run at least for me.

Cleaned up #31.
Added leases, config, more method stubs, and then we can get rid of the god forsaken rpc-impl branch.

### CODE REVIEW

File: rice-namenode/ClientNamenodeProtocolImpl.cc [hunk 1]
```diff
@@ -24,8 +28,7 @@ namespace client_namenode_translator {
 // the .proto file implementation's namespace, used for messages
 using namespace hadoop::hdfs;
 
-// static string info
-const char* ClientNamenodeTranslator::HDFS_DEFAULTS_CONFIG = "hdfs-default.xml";
+const int ClientNamenodeTranslator::LEASE_CHECK_TIME = 60;
```
Comment (pelmers) [inline rice-namenode/ClientNamenodeProtocolImpl.cc]: should comment the units here, or put it in the variable name (e.g. lease_check_seconds)

File: rice-namenode/ClientNamenodeProtocolImpl.cc [hunk 2]
```diff
@@ -34,148 +37,263 @@ std::map <std::string, std::string> config;
...
+	CompleteRequestProto req;
+	req.ParseFromString(input);
+	logMessage(req, "Complete ");
+	const std::string& src = req.src();
+	const std::string& clientname = req.clientname();
+	// TODO some optional fields need to be read
+	// remove the lease from this file  
+	bool succ = lease_manager.removeLease(clientname, src);         
+	if (!succ) {
+		LOG(ERROR) << "A client tried to close a file which is not theres";
```
Comment (pelmers) [inline rice-namenode/ClientNamenodeProtocolImpl.cc]: little typo, theres

File: rice-namenode/ClientNamenodeProtocolImpl.cc [hunk 3]
```diff
@@ -34,148 +37,263 @@ std::map <std::string, std::string> config;
...
+std::string ClientNamenodeTranslator::abandonBlock(std::string input) {
+	AbandonBlockRequestProto req;
+	req.ParseFromString(input);
+	logMessage(req, "AbandonBlock ");
+	const ExtendedBlockProto& blockProto = req.b();
+	const std::string& src = req.src();
+	const std::string& holder = req.src(); // TODO who is the holder??
+	// TODO some optional fields
+	// TODO tell zookeeper to get rid of the last block (should not need to 
+    // (talk to datanode as far as i am aware)
```
Comment (pelmers) [inline rice-namenode/ClientNamenodeProtocolImpl.cc]: fricking tabs

File: rice-namenode/ClientNamenodeProtocolImpl.h [hunk 4]
```diff
@@ -17,28 +19,44 @@ using namespace hadoop::hdfs;
...
 		void RegisterClientRPCHandlers();
-		void Config();
-		void logMessage(google::protobuf::Message& req);
+		void logMessage(google::protobuf::Message& req, std::string req_name);
+		void leaseCheck();
+		
+		int getDefaultInt(std::string);
 
-		FsServerDefaultsProto server_defaults;
+        static const int LEASE_CHECK_TIME;
```
Comment (pelmers) [inline rice-namenode/ClientNamenodeProtocolImpl.h]: fricking tabs

File: rice-namenode/Leases.cc [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+ * This happens every LEASE_CHECK_TIME seconds. If the file is
+ * expired, then we forcibly close it. Otherwise, we hadd 60
+ * seconds onto the timer 
+ */
+bool Lease::isExpired(int time) {
+	// TODO close the file and recover the blocks
+	if (time >= LEASE_EXPIRATION_TIME) {
+		return false;
+	}
+	time += time;
```
Comment (pelmers) [inline rice-namenode/Leases.cc]: the scoping always gets me turned around. can we rename the local variable so it doesn't look like the class variable?

**Review by pelmers — COMMENTED** [review]
I appreciate the refactoring on logMessage and Serialize!
(remark: 'fricking tabs' comment means indentation)

r+

File: rice-namenode/ClientNamenodeProtocolImpl.cc [hunk 6]
```diff
@@ -186,21 +304,28 @@ void ClientNamenodeTranslator::InitServer() {
...
 	// The reason for these binds is because it wants static functions, but we want to give it member functions
     // http://stackoverflow.com/questions/14189440/c-class-member-callback-simple-examples
-
 	server.register_handler("getFileInfo", std::bind(&ClientNamenodeTranslator::getFileInfo, this, _1));
 	server.register_handler("mkdir", std::bind(&ClientNamenodeTranslator::mkdir, this, _1));
 	server.register_handler("append", std::bind(&ClientNamenodeTranslator::append, this, _1));
 	server.register_handler("destroy", std::bind(&ClientNamenodeTranslator::destroy, this, _1));
 	server.register_handler("create", std::bind(&ClientNamenodeTranslator::create, this, _1));
 	server.register_handler("getBlockLocations", std::bind(&ClientNamenodeTranslator::getBlockLocations, this, _1));
+
```
Comment (pelmers) [inline rice-namenode/ClientNamenodeProtocolImpl.cc:327]: are there more handlers to register here?
here's the methods defined in the translator but not registered:
getServerDefaults
renewLease
complete
setReplication
recoverLease
concat
abandonBlock

**Comment by kyler-m** [comment]
r+

**Pull request merged**